### ISSUE TITLE
Compatibility with v9 .hic files

### ISSUE THREAD

**Issue by sa501428** [vaquerizaslab/fanc#92]
Hi,

We've had some users reach out about using your tool with the latest version of .hic files.
Is this something that could be rectified fairly easily?

We have the straw library which can directly read .hic data into python: https://github.com/aidenlab/straw
Alternatively, the jar used by fanc could be updated to the latest juicer_tools jar so that the newer version of dump is used.

Forum discussion: https://mail.google.com/mail/u/0/#inbox/FMfcgzGmtXMhNktCkCQTFvgnvqPKBNfh

**Comment by kaukrise** [comment]
Hi, thanks a lot for the info!

The link you provided does not work for me, but I suppose it is this discussion thread? https://groups.google.com/g/3d-genomics/c/msQn1MZfF7I

Would you have a small sample `v9` file for me to work with? The Encode file is ~50Gb large and the estimated download time is >3 hours with my connection...

FAN-C uses its own compatibility code for Juicer files and does not use the `jar` at all. The compatibility layer was originally inspired by straw, but back when I started working on it, straw was fairly rudimentary. Looks like that has changed - it looks awesome now! I'll try to fix the v9 compatibility in the FAN-C code first, and then I'll look into substituting this for straw down the line.

Cheers,

Kai

**Comment by kaukrise** [comment]
Hi again, I was able to generate my own Juicer `v9` sample file for testing using `juicer-tools` `v2.13.07`.

I made some changes to the Juicer compatibility layer using the v9-specific changes I could identify in the `straw` [C++ code](https://github.com/aidenlab/straw/blob/master/C%2B%2B/straw.cpp), which was very helpful.

The changes appear to work on the sample file and the ENCODE file `ENCFF053VBX.hic`. I tried to reply to the forum thread, too, but I do not seem to have permission?

Anyways, if you could refer the user here, please, maybe they can test it out on their files. The beta version with the compatibility fix can be installed from this tarball
[fanc-0.9.23b0.tar.gz](https://github.com/vaquerizaslab/fanc/files/7933961/fanc-0.9.23b0.tar.gz)

```
pip install fanc-0.9.23b0.tar.gz
```

**Comment by sa501428** [comment]
Oh awesome! 
Here's another test file as well: https://www.dropbox.com/s/g728zbhrwtndn0q/test.hic

We also deprecated KR in favor of SCALE, which converges more reliably and normalizes the matrices faster. For a full list of changes to the format, see: https://github.com/aidenlab/HICV9/blob/master/HiCFormatV9.md

And my bad, I shared the wrong URL. Yes, that is the correct post. If you request to join the group as a member, I can approve you / only members can post due to lots of recent spam on the Google Group. Would be great if you can post there. Or I can point the user here as well. Lmk.

Thanks so much!

**Comment by chongli0710** [comment]
> Hi again, I was able to generate my own Juicer `v9` sample file for testing using `juicer-tools` `v2.13.07`.
> 
> I made some changes to the Juicer compatibility layer using the v9-specific changes I could identify in the `straw` [C++ code](https://github.com/aidenlab/straw/blob/master/C%2B%2B/straw.cpp), which was very helpful.
> 
> The changes appear to work on the sample file and the ENCODE file `ENCFF053VBX.hic`. I tried to reply to the forum thread, too, but I do not seem to have permission?
> 
> Anyways, if you could refer the user here, please, maybe they can test it out on their files. The beta version with the compatibility fix can be installed from this tarball [fanc-0.9.23b0.tar.gz](https://github.com/vaquerizaslab/fanc/files/7933961/fanc-0.9.23b0.tar.gz)
> 
> ```
> pip install fanc-0.9.23b0.tar.gz
> ```

Hi Kai,

Thank you for the update on such short time! As you mentioned, the beta version worked on the ENCODE file ENCFF053VBX.hic, which is the same file I have tested on. But the fan-c returned me error as " ValueError: Cannot find normalisation 'KR' ". Have you met the same error from your side? 

Here is my code: insulation = fanc.InsulationScores.from_hic(hic_5kb,
                                            [100000],
                                            file_name="fanc_insulation/inter30.insulation")

Thanks!

**Comment by sa501428** [comment]
Is there a way to request "SCALE" instead of "KR" in the parameters?

**Comment by kaukrise** [comment]
Can you try loading the `hic` file with

```
hic_5kb = fanc.load("/path/to/file.hic@5000@SCALE")
```

please?

**Comment by kaukrise** [comment]
And thanks a lot, @sa501428 , the added detail is much appreciated. I'll go over it when I have a bit more time

**Comment by chongli0710** [comment]
Yes, by default, fan-c retrieves KR norm and I just manually changed it to SCALE and it seems working now. Thank you Muhammad for helping raise this issue to fan-c team. May I also confirm with you that does the new version of .hic file no longer provide KR normalization? Since I used to generate several .hic files for my samples by doing the KR norm and I am wondering should I redo them all to apply SCALE norm to keep consistent with the ENCODE .hic file. Have you published any papers/manuscript about why SCALE works better than KR? Thank you again! 发件人: Muhammad Saad Shamim发送时间: 2022年1月25日 13:54收件人: vaquerizaslab/fanc抄送: chongli0710; Comment主题: Re: [vaquerizaslab/fanc] Compatibility with v9 .hic files (Issue #92) Is there a way to request "SCALE" instead of "KR" in the parameters?—Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android. You are receiving this because you commented.Message ID: ***@***.***>

**Comment by chongli0710** [comment]
Just did the same thing as you suggest, worked perfectly now. Many thanks Kai, I really like fan-c! 发件人: kaukrise发送时间: 2022年1月25日 13:58收件人: vaquerizaslab/fanc抄送: chongli0710; Comment主题: Re: [vaquerizaslab/fanc] Compatibility with v9 .hic files (Issue #92) Can you try loading the hic file withhic_5kb = ***@***.***@SCALE")please?—Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android. You are receiving this because you commented.Message ID: ***@***.***>

**Comment by kaukrise** [comment]
Hey @sa501428 , I was playing around with strawC, and since I am not familiar with C++ bindings in Python I thought I'd ask you directly:

The FAN-C compatibility layer needs access to specific elements of the Hi-C map, and I think I can find matching functions in the C++ code for each of them:

- Available resolutions and normalisations (`readResolutionsFromHeader`)
- The chromosome map (`readHeader`)
- Expected value vectors (`readFooter` ?)
- Normalisation vectors (`readNormalizationVector` in combination with `readFooter`)
- The raw contact matrix (` readMatrix`)

However, I do not see those functions exposed in the `strawC` API, only `strawC.strawC`, which returns a `contactRecord` with bin and contact information. Is there any way to get access to the above individual elements of the Hi-C matrix from the API?

Thanks a lot in advance!

**Comment by sa501428** [comment]
We should have all of these resolved with the latest updates (not yet merged into master, but hopefully by the end of the week)
https://github.com/sa501428/straw/tree/helper_methods

**Comment by kaukrise** [comment]
Oooh, that's amazing, I'll install from source then to try it out! Thanks for the fast response!

**Comment by sa501428** [comment]
No problem! There may still be some bugs / we're still testing - so let us know if you encounter any issues!

**Comment by sa501428** [comment]
We also need to update the usage / will send this along.

**Comment by rdacemel** [comment]
Hi! 

I'm experiencing some weird issues after I switch to .hic v9. When it comes to plotting, some coords appear just white. It does not make much sense because, as you can see in the image, those coordinates can be plotted if some padding is added. I'm really confused, and when inspecting the region in Juicebox nothing weird pops up. 

![image](https://user-images.githubusercontent.com/14839665/152213478-7da217cb-073f-4e78-b711-5f9416fdd9f8.png)

Maybe it's a matter of accessing the normalizations? Is there a way to know how many normalizations are included inside a fanc-juicer object?

**Comment by kaukrise** [comment]
Hi, thanks for reporting this, it is indeed odd. Can you please try a couple of things?

- Could you plot the same region using `fancplot` from the command line? Are there any logging messages or warnings?
- Can you try to plot the same region without normalisation? `fanc.load("/path/to/inter_30.hic@25kb@NONE")`
- Can you paste the bias vector for this region? `print([r.bias for r in hic.regions("chr17:30mb-35mb")])`

It looks from the image as though the matrix is masked, which indeed indicates some issue with the normalisation.

**Comment by kaukrise** [comment]
Hey, forget what I wrote above, I was able to reproduce the issue on my own!

Looks like I missed a crucial piece of new `v9` file structure, apologies for that. I think I found and corrected the issue - could you please test it on your data and let me know if this also works for you?

[fanc-0.9.23b2.tar.gz](https://github.com/vaquerizaslab/fanc/files/7989577/fanc-0.9.23b2.tar.gz)

To be safe, I would also recalculate the insulation score using this version if you have no further issues.

Thanks!
Kai

**Comment by rdacemel** [comment]
It's working now! Thanks, ultra-fast as always. It is writing some numbers now though ;)

**Comment by kaukrise** [comment]
Yes, sorry, the number are just debug messages I thought I might need in case things still don't work. Here is a version without debug messages. This version should also select 'SCALE' for v9 files by default and warn that v9 support is still in beta.

[fanc-0.9.23b3.tar.gz](https://github.com/vaquerizaslab/fanc/files/7993295/fanc-0.9.23b3.tar.gz)

**Comment by rdacemel** [comment]
Nice! Thank you so much!
